hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java

```diff
@@ -856,13 +856,9 @@ protected void rollbackFailedWrites(List<String> instantsToRollback, boolean ski
       } else {
         rollback(instant, skipLocking);
       }
-    }
-    // Delete any heartbeat files for already rolled back commits
-    try {
-      HeartbeatUtils.cleanExpiredHeartbeats(this.heartbeatClient.getAllExistingHeartbeatInstants(),
-          createMetaClient(true), basePath);
-    } catch (IOException io) {
-      LOG.error("Unable to delete heartbeat files", io);
+
+      // Delete the heartbeat for rolled back instant
+      HeartbeatUtils.deleteHeartbeatFile(fs, basePath, instant, config);
```

Comment:
In this patch, we delete the rollback instant in the `for (String instant : instantsToRollback)` loop. 
Is there may something un-deleted when going into https://github.com/apache/hudi/blob/add8acecb367ee7b8506e5abda9137d203e34645/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java#L855 branch ?

Maybe we could do the delete HeartbeatsFiles out of the loop :)

Reply:
Moving it outside the loop has the risk that if there are multiple commits to rollback, and if for some reason rollback succeeds for earlier commits but fails for a commit later in the list, then the job would not clear the heartbeat files even for commits that were rolled back correctly. Its best to also clear the heartbeat as soon as the commit is rolled back, so other commits do not have any affect on this. Also in terms of just time, moving it outside the loop could create a time lag between when a commit is rolled back and when its heartbeat file is deleted.